### DISCUSSION TITLE
any performance docs

### DISCUSSION THREAD

**Discussion by Courvoisier13** [pwwang/datar#132] [General]
this is brilliant, please keep it up.

Do you have any data on performance?

**Comment by pwwang** [comment]
Nope for now.

There are some concerns about benchmarking. Most of the verbs or functions can not be implemented with a simple pandas equivalent, which makes it arbitrary for the counterpart implementations, resulting in unfair comparisons.

When comparing this library with other similar libraries. It also comes up with unfair comparisons, because `datar` does not only mimic the tidyverse APIs, but also implements them side-by-side using their source code from tidyverse as references. This way, it can handle some "complex" scenarios with `datar` APIs.

Some "complex" examples can be found here:

- https://stackoverflow.com/questions/71518997/count-by-id-with-dynamic-criteria/71519157#71519157
- https://stackoverflow.com/questions/69257878/replicate-r-solution-in-python-for-calculating-monthly-crr/71490194#71490194
- https://stackoverflow.com/questions/71165372/how-to-transform-r-dataframe-to-rows-of-indicator-values/71443515#71443515

You may find more on SO. I don't think the current similar tools support these features. 
The philosophy of `datar` is trying to make `datar` more aligned with tidyverse APIs, even with complex expressions. However, it's definitely sacrificing some performance to support this.

Another example you can find is from here:

#128 

You can see that when we want to use more features, the performance is close to pandas implementation. But when we just want to do a simple mutation, `datar` `mutate` is defeated, as it considers more complicated situations. Of course, this can be optimized by distinguishing simple and complicated cases through further efforts.

Open to ideas and proposals!
